### PR TITLE
Add a modify function for rects

### PR DESCRIPTION
**Pull request by tokenrove**
Without a way to modify rects, programs that use them in must cons new
ones every time.  With no option to give them dynamic extent, this
can create a huge amount of garbage.

I needed to create this modification to get Zookicker (http://github.com/tokenrove/zookicker) running, a trivial game using tsdl, but I appreciate that it is probably not the most elegant approach.

I would appreciate feedback on how this goal could be achieved in a way that fits tsdl's design.

### CODE REVIEW

**Comment by dbuenzli** [comment]
Somehow I was hoping that basic structs could remain immutable. Did you really get into performance problems because of this ? 

In any case if we add this it should be done consistently for the other base structs.

File: src/tsdl.mli [hunk 1]
```diff
@@ -353,6 +353,7 @@ type rect
 
 module Rect : sig
   val create : x:int -> y:int -> w:int -> h:int -> rect
+  val modify : rect -> x:int -> y:int -> w:int -> h:int -> rect
```
Comment (dbuenzli) [inline src/tsdl.mli]: Regardless of acceptance this should return unit, the signature gives a wrong idea of what it is doing.

**Comment by dbuenzli** [comment]
Maybe it makes sense if you are using SDL's 2D rendering engine, I can see the idea of having your own OCaml data structures and wanting only a few allocated rects to interface with SDL's 2D drawing functions. For example that's what the interface to the event API allows you to do: you can use a single event struct for handling all your events, so that higher-level abstractions can be constructed on top of that without having to allocate too much at the C interface level.

I just had a look at the current API I think that if we want this then it should be consistantly done on  `Rect`, `Color` and `Point` and rather than provide a bulk function like this PR proposes, add single component setters mirroring the accessors. This would mean adding these functions: 

```ocaml
Color.set_{r,g,b,a} : color : uint8 -> unit
Point.set_{x,y} : point : int -> unit
Rect.set_{x,y,w,h} : rect -> int -> unit 
```

File: src/tsdl.mli (hunk 1 above)
Comment (tokenrove) [inline src/tsdl.mli]: Oops.  Definitely.

**Comment by tokenrove** [comment]
The performance impact is massive even with a handful of rects being used per frame.

Other FFI systems usually provide a way to stack allocate a temporarily-used foreign structure; it's too bad that's not an option here.

**Comment by dbuenzli** [comment]
> The performance impact is massive even with a handful of rects being used per frame.

Ok, I never saw that since never used SDL's graphics engine, I only used `tsdl` with OpenGL. Also I'm not sure how exactly ctypes manages the memory allocated by `Ctypes.make`.

**Comment by tokenrove** [comment]
Yeah, it's funny, but if I hadn't been trying to throw this thing together quickly, I wouldn't have used SDL's renderer stuff, and I wouldn't have noticed this problem.

I'll add those setters and submit a new pull request.  Thanks.

**Pull request closed without merging**